org.eclipse.xtext.builder.tests/src/org/eclipse/xtext/builder/impl/AbstractFSSynchronizationTest.java

```diff
@@ -0,0 +1,427 @@
+		});
+	}
+
+	@Test
+	public void testCleanProjectAsOutput() {
+		withOutputConfiguration((OutputConfiguration it) -> {
+			it.setOutputDirectory("./");
+			it.setCanClearOutputDirectory(true);
+			it.setCleanUpDerivedResources(false);
+		}, () -> {
+			File projectDirectory = project.getLocation().toFile();
+			int initialSize = projectDirectory.listFiles().length;
+			Assert.assertNotEquals(0, initialSize);
+			createFile(project.getFile(("src/Foo" + F_EXT)).getFullPath(), "object Foo");
+			build();
+			int expectedSize = projectDirectory.listFiles().length;
+			Assert.assertNotEquals(initialSize, expectedSize);
+			cleanBuild();
+			Assert.assertEquals(expectedSize, projectDirectory.listFiles().length);
+		});
+	}
+
+	@Test
+	public void testCleanUpDerivedResourcesWithCreateBefore() {
+		testCleanUpDerivedResourcesWithCreateBefore(project.getFolder("src-gen"));
+	}
+
+	@Test
+	public void testCleanUpDerivedResourcesWithCreateBetween() {
+		testCleanUpDerivedResourcesWithCreateBetween(project.getFolder("src-gen"));
+	}
+
+	@Test
+	public void testCleanUpDerivedResourcesWithUpdateDerived() {
+		testCleanUpDerivedResourcesWithUpdateDerived(project.getFolder("src-gen"));
+	}
+
+	@Test
+	public void testCleanUpDerivedResourcesWithDeleteDerived() {
+		testCleanUpDerivedResourcesWithDeleteDerived(project.getFolder("src-gen"));
+	}
+
+	@Test
+	public void testCleanUpProjectDerivedResourcesWithCreateBefore() {
+		withOutputConfiguration(it -> it.setOutputDirectory("./"), () -> testCleanUpDerivedResourcesWithCreateBefore(project));
+	}
+
+	@Test
+	public void testCleanUpProjectDerivedResourcesWithCreateBetween() {
+		withOutputConfiguration(it -> it.setOutputDirectory("./"), () -> testCleanUpDerivedResourcesWithCreateBetween(project));
+	}
+
+	@Test
+	public void testCleanUpProjectDerivedResourcesWithUpdateDerived() {
+		withOutputConfiguration(it -> it.setOutputDirectory("./"), () -> testCleanUpDerivedResourcesWithUpdateDerived(project));
+	}
+
+	@Test
+	public void testCleanUpProjectDerivedResourcesWithDeleteDerived() {
+		withOutputConfiguration(it -> it.setOutputDirectory("./"), () -> testCleanUpDerivedResourcesWithDeleteDerived(project));
+	}
+
+	protected void testCleanUpDerivedResourcesWithCreateBefore(IContainer output) {
+		File ouputDirectory = createJavaIoFile(output.getFile(new Path("Lalala.txt")).getLocation()).getParentFile();
+		int expectedSize = ouputDirectory.list().length;
+		createFile(project.getFile(("src/Foo" + F_EXT)).getFullPath(), "object Foo");
+		build();
+		Assert.assertNotEquals(expectedSize, ouputDirectory.list().length);
+		disableAutobuild(() -> {
+			cleanBuild();
+			Assert.assertEquals(expectedSize, ouputDirectory.list().length);
+			Assert.assertTrue(ArrayExtensions.contains(ouputDirectory.list(), "Lalala.txt"));
+		});
+	}
+
+	protected void testCleanUpDerivedResourcesWithCreateBetween(IContainer output) {
+		File outputDirectory = output.getLocation().toFile();
+		final int initialSize;
```

Comment:
Some unnecessary final locals in this class, but again not crucial.

Reply:
Local variable initialSize defined in an enclosing scope must be final or effectively final